CI, Test coverage

zrock-dev opened the pull request:
# Task 150| Test Coverage

https://tree.taiga.io/project/joseluis-teran-coffeetime/task/150

## Description

Provide a small description of Pull Request or answer the following questions if it is necessary.

#### What did I do?

Added a test coverage analyst to the pull-request CI

#### How did I do it?

- added JaCoCo test coverage tool, It's been configured to run after all tests pass, it generates an xml report per test class
- added [codecov](https://github.com/codecov/codecov-action) to report actions

#### Why did I do it?

We need to ensure a stable library

## Type of Change

Put an `x` in all the boxes that apply:

- [ ] ✨ New Feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug Fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking Change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code Refactor
- [x] ✅ Build Configuration Change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

## Pull Request Checklist

Put an `x` in all the boxes that apply:

- [x] My commit messages are detailed.
- [x] My code follows the code style of this project.
- [x] No existing features have been broken without good reason.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests pass.

zrock-dev committed:
chore: added JaCoCo test coverage tool

- It's been configured to run after all tests pass, it generates an xml
  report per test class

zrock-dev committed:
chore: added codecov to report actions

zrock-dev committed:
refactor: build job without tests

- The build job shouldn't perform test tasks that's what the test job is
  for

zrock-dev committed:
doc: documented JaCoCo in the README

- Added instructions on how to generate reports and preview them.

teran-joseluis commented on .github/workflows/check-pull-request.yml:
```diff
@@ -37,6 +37,24 @@ jobs:
+        run: |
+          ./gradlew --no-daemon test jacocoTestReport -x assemble
+
+      - name: Publish coverage reports
```
Can you update the name to be more descriptive?

zrock-dev replied:
like what? publish coverage reports to codecov?

teran-joseluis replied:
Yeah, I'd be better.

zrock-dev replied:
haha, it's just like the same name bruh

teran-joseluis replied:
Well, it's up to you.

teran-joseluis commented on .github/workflows/check-pull-request.yml:
```diff
@@ -37,6 +37,24 @@ jobs:
+        uses: codecov/codecov-action@v4
+        with:
+          fail_ci_if_error: true 
+          files: ./bridge/build/reports/jacoco/test/jacocoTestReport.xml
```
The path you have entered is not generating the report property, check it. 
However it generates the report here.
![image](https://github.com/Pending-Name-21/frontend-library/assets/73414963/627c8d90-814c-4c33-926f-c417e23f5e74)

teran-joseluis commented on .github/workflows/check-pull-request.yml:
```diff
@@ -37,6 +37,24 @@ jobs:
+        uses: codecov/codecov-action@v4
+        with:
+          fail_ci_if_error: true 
+          files: ./bridge/build/reports/jacoco/test/jacocoTestReport.xml
```
I see the extension let us to define if we want a file or list of files for the reports. Can you explain me why do you choose files instead of file?

zrock-dev replied:
It's because codecov requires a specific file in order to generate the report, which is this .xml. And the CI it's explicitly generating this during the  `jacocoTestReport` task

teran-joseluis replied:
Yeah, I see your point but you didn't catch what I'm saying. When you pick files option, you can generate more than one .xml file, but in this case we are only generating one so I'd be better to use file option.

zrock-dev replied:
oh I get it now

teran-joseluis replied:
Nice

teran-joseluis commented on .github/workflows/check-pull-request.yml:
```diff
@@ -37,6 +37,24 @@ jobs:
+          token: ${{ secrets.CODECOV_TOKEN }}
+          verbose: true
+          plugin: noop
 
```
Also, don't you think it will be good the specify the OS we are using to run it? I see the extension let us to put it.
https://github.com/marketplace/actions/codecov

zrock-dev replied:
Okay, will do

teran-joseluis replied:
Nice

teran-joseluis commented on .github/workflows/check-pull-request.yml:
```diff
@@ -37,6 +37,24 @@ jobs:
+          token: ${{ secrets.CODECOV_TOKEN }}
+          verbose: true
+          plugin: noop
 
```
Remove the extra empty line you added.

zrock-dev replied:
> It's not about adding an extra newline at the end of a file, it's about not removing the newline that should be there.
A [text file](http://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap03.html#tag_03_392), under unix, consists of a series of [lines](http://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap03.html#tag_03_205), each of which ends with a [newline character](http://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap03.html#tag_03_238) (\n). A file that is not empty and does not end with a newline is therefore not a text file.

-- [source](https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file)

teran-joseluis replied:
I know it, but you didn't understand what I was saying.
In `check-push.yml` file we have the newline character, which is good and refers to what you mention. But in the other one, `check-pull-request.yml`, we have two newline characters. 
The images of both files are shown below.

`check-push.yml`
![image](https://github.com/Pending-Name-21/frontend-library/assets/73414963/76bcee50-f122-4dd0-9e9f-8b9a60cd48f2)

`check-pull-request.yml`
![image](https://github.com/Pending-Name-21/frontend-library/assets/73414963/81e78e58-9127-4902-afa2-119d82f3e169)

zrock-dev replied:
Are you sure about that? the git diff ony shows one...

`$ git diff --staged --patch origin/main -- .github/workflows/check-pull-request.yml  | tail`

![image](https://github.com/Pending-Name-21/frontend-library/assets/64939916/4c1efb07-253f-46f6-8215-3899641ddf3a)

teran-joseluis replied:
Did you see the difference?
![image](https://github.com/Pending-Name-21/frontend-library/assets/73414963/e98207c0-931c-4720-a896-54c19475c972)

teran-joseluis replied:
Did you see the difference between both?
![image](https://github.com/Pending-Name-21/frontend-library/assets/73414963/e9e4b7af-0af8-45bc-bfee-2d0f8b2a8042)

zrock-dev replied:
Yeah, about that... the context of this PR is just the pull-request yml.

teran-joseluis replied:
I was not talking about the context of the PR, I mentioned that file in order to see extra empty line you added besides of required one.

zrock-dev replied:
bruh, I'm telling you, there are no extra lines. It has 62 lines, the 62nd line is the empty line

teran-joseluis replied:
Are you sure? Why don't you review it using an editor?

teran-joseluis replied:
No changes:
![image](https://github.com/Pending-Name-21/frontend-library/assets/73414963/123677cb-94c8-4247-b2ed-f665defac336)

Deleting the extra line you cannot see.
![image](https://github.com/Pending-Name-21/frontend-library/assets/73414963/ebcd313e-9f4f-414e-b8a8-bc91046914ff)

teran-joseluis replied:
No changes:
![image](https://github.com/Pending-Name-21/frontend-library/assets/73414963/cc400292-6aa1-4873-9134-52da4c1a3662)

Deleting the extra newline you cannot see.
![image](https://github.com/Pending-Name-21/frontend-library/assets/73414963/3d9897ea-cc0f-4127-bf4e-7e61e57c6c70)

zrock-dev replied:
I've opened with an editor as well.

can you try using this command?

`$  git show origin/task-150/test-coverage:.github/workflows/check-pull-request.yml | wc -l`

it should print 62, the amount of lines that file has

teran-joseluis commented on README.md:
```diff
@@ -27,3 +27,12 @@ Then you can add bridge as any other dependency.
+` $./gradlew build
+` $./gradlew test jacocoTestReport
+
+JaCoCo generates HTML reports which canb be found at "./bridge/build/reports/jacoco/test/html"
```
Update the path of the report.

teran-joseluis commented on README.md:
```diff
@@ -27,3 +27,12 @@ Then you can add bridge as any other dependency.
+` $./gradlew build
+` $./gradlew test jacocoTestReport
+
+JaCoCo generates HTML reports which canb be found at "./bridge/build/reports/jacoco/test/html"
```
Update the path of the report according the ./gradlew task executed please so that others can guide themselves.

zrock-dev replied:
that's the path from project root

teran-joseluis replied:
Yeah, but when it's performed by build task, it generates a different path for the report which is not the same you wrote it.

zrock-dev replied:
Are you sure about that? I've removed the older build and generated a new one with`$ ./gradlew test jacocoTestReport` and the path is there.

teran-joseluis replied:
I agree with you with you on this one `$ ./gradlew test jacocoTestReport`, but on `$ ./gradlew build` task the path is different.
![image](https://github.com/Pending-Name-21/frontend-library/assets/73414963/e038d9dd-15f2-4a11-abd7-871fa41c907f)

My suggestion is to update something like this:

Test coverage can be generated either
` $./gradlew build` 
`$./gradlew test jacocoTestReport`

Using build task the path is "./bridge/build/reports/tests/test/index.html"
JaCoCo generates HTML reports which **canb** be found at "./bridge/build/reports/jacoco/test/html"

Also, update the canb word to can please.

zrock-dev replied:
oke

teran-joseluis commented on .github/workflows/check-pull-request.yml:
```diff
@@ -37,6 +37,24 @@ jobs:
+          token: ${{ secrets.CODECOV_TOKEN }}
+          verbose: true
+          plugin: noop
 
```
Also, can you add flags in order to difference the coverage metrics?
https://github.com/marketplace/actions/codecov

zrock-dev replied:
The [flags feature](https://about.codecov.io/product/feature/flags/) it's targeted mainly for mono-repo codebases. Or if it has different kind of tests, but we only have unit testing. I don't think is worth it
![image](https://github.com/Pending-Name-21/frontend-library/assets/64939916/51c6353f-b9a4-48d4-96ec-e9203ece59a1)

teran-joseluis replied:
I know that we only have unit testing but when the Codecov Report is generated it says service and not unit testing, that's why I asked to you to put it.
![image](https://github.com/Pending-Name-21/frontend-library/assets/73414963/d91b08e0-8db2-4b86-bb70-237a2fcfd19a)

zrock-dev replied:
Sure

teran-joseluis replied:
Nice

teran-joseluis replied:
I see you added our flag but I have a question, is the flag service generated by default?

zrock-dev replied:
The [GitHub action](https://github.com/codecov/codecov-action) does that by default anyway you can read the [docs](https://docs.codecov.com/docs/flags)
![image](https://github.com/Pending-Name-21/frontend-library/assets/64939916/169c3b33-3f37-4004-aefb-89c9093b35f4)

teran-joseluis replied:
Okay!

zrock-dev commented:
I have the commit ready, but I'm waiting to finish the other conversations before pushing

zrock-dev committed:
doc: updated test report to be more descriptive

I've changed the description of the generation paths. Given that both
tasks test and build generate test reports.

zrock-dev committed:
chore: updated codecov configuration

- Added metadata like os and flags to have more information about the
  tests.
- refactor files to file since there is only one file being used

teran-joseluis commented:
> I have the commit ready, but I'm waiting to finish the other conversations before pushing

There is one more open and it's done.

teran-joseluis commented:
> > I have the commit ready, but I'm waiting to finish the other conversations before pushing
> 
> There is one more open and it's done.

Don't forget to mark as resolved please.

teran-joseluis commented:
> I have the commit ready, but I'm waiting to finish the other conversations before pushing

I had written some comments more, please review it.

teran-joseluis reviewed (approved):
Good job!
